File: app/src/main/java/dev/ragnarok/fenrir/db/impl/PhotosStorage.java
package dev.ragnarok.fenrir.db.impl;

import static dev.ragnarok.fenrir.util.Objects.nonNull;
import static dev.ragnarok.fenrir.util.Utils.nonEmpty;
import static dev.ragnarok.fenrir.util.Utils.safeCountOf;

import android.content.ContentProviderOperation;
import android.content.ContentValues;
import android.database.Cursor;
import android.net.Uri;
import android.provider.BaseColumns;

import androidx.annotation.NonNull;

import java.util.ArrayList;
import java.util.List;

import dev.ragnarok.fenrir.db.DatabaseIdRange;
import dev.ragnarok.fenrir.db.MessengerContentProvider;
import dev.ragnarok.fenrir.db.column.PhotosColumns;
import dev.ragnarok.fenrir.db.interfaces.IPhotosStorage;
import dev.ragnarok.fenrir.db.model.PhotoPatch;
import dev.ragnarok.fenrir.db.model.entity.PhotoEntity;
import dev.ragnarok.fenrir.db.model.entity.PhotoSizeEntity;
import dev.ragnarok.fenrir.model.criteria.PhotoCriteria;
import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Single;

class PhotosStorage extends AbsStorage implements IPhotosStorage {

    PhotosStorage(@NonNull AppStorages base) {
        super(base);
    }

    private static ContentValues getCV(PhotoEntity dbo) {
        ContentValues cv = new ContentValues();
        cv.put(PhotosColumns.PHOTO_ID, dbo.getId());
        cv.put(PhotosColumns.ALBUM_ID, dbo.getAlbumId());
        cv.put(PhotosColumns.OWNER_ID, dbo.getOwnerId());
        cv.put(PhotosColumns.WIDTH, dbo.getWidth());
        cv.put(PhotosColumns.HEIGHT, dbo.getHeight());
        cv.put(PhotosColumns.TEXT, dbo.getText());
        cv.put(PhotosColumns.DATE, dbo.getDate());

        if (nonNull(dbo.getSizes())) {
            cv.put(PhotosColumns.SIZES, GSON.toJson(dbo.getSizes()));
        }

        cv.put(PhotosColumns.USER_LIKES, dbo.isUserLikes());
        cv.put(PhotosColumns.CAN_COMMENT, dbo.isCanComment());
        cv.put(PhotosColumns.LIKES, dbo.getLikesCount());
        cv.put(PhotosColumns.COMMENTS, dbo.getCommentsCount());
        cv.put(PhotosColumns.TAGS, dbo.getTagsCount());
        cv.put(PhotosColumns.ACCESS_KEY, dbo.getAccessKey());
        cv.put(PhotosColumns.DELETED, dbo.isDeleted());
        return cv;
    }

    private static String getSelectionForCriteria(@NonNull PhotoCriteria criteria) {
        String selection = "1 = 1";
        if (criteria.getOwnerId() != null) {
            selection = selection + " AND " + PhotosColumns.OWNER_ID + " = " + criteria.getOwnerId();
        }

        if (criteria.getAlbumId() != null) {
            selection = selection + " AND " + PhotosColumns.ALBUM_ID + " = " + criteria.getAlbumId();
        }

        DatabaseIdRange range = criteria.getRange();
        if (nonNull(range)) {
            selection = selection + " AND " + BaseColumns._ID + " >= " + range.getFirst() +
                    " AND " + BaseColumns._ID + " <= " + criteria.getRange().getLast();
        }

        return selection;
    }

    @Override
    public Completable insertPhotosRx(int accountId, int ownerId, int albumId, @NonNull List<PhotoEntity> photos, boolean clearBeforeInsert) {
        return Completable.fromAction(() -> {
            ArrayList<ContentProviderOperation> operations = new ArrayList<>(clearBeforeInsert ? photos.size() + 1 : photos.size());
            Uri uri = MessengerContentProvider.getPhotosContentUriFor(accountId);

            if (clearBeforeInsert) {
                operations.add(ContentProviderOperation
                        .newDelete(uri)
                        .withSelection(PhotosColumns.OWNER_ID + " = ? AND " + PhotosColumns.ALBUM_ID + " = ?"
                                , new String[]{String.valueOf(ownerId), String.valueOf(albumId)})
                        .build());
            }

            for (PhotoEntity dbo : photos) {
                operations.add(ContentProviderOperation
                        .newInsert(uri)
                        .withValues(getCV(dbo))
                        .build());
            }

            getContext().getContentResolver().applyBatch(MessengerContentProvider.AUTHORITY, operations);
        });
    }

    @Override
    public Single<List<PhotoEntity>> findPhotosByCriteriaRx(@NonNull PhotoCriteria criteria) {
        return Single.create(e -> {
            String selection = getSelectionForCriteria(criteria);

            String orderBy = criteria.getOrderBy() == null ? PhotosColumns.PHOTO_ID + (!criteria.getSortInvert() ? " DESC" : " ASC") : criteria.getOrderBy();

            Uri uri = MessengerContentProvider.getPhotosContentUriFor(criteria.getAccountId());
            Cursor cursor = getContext().getContentResolver().query(uri, null, selection, null, orderBy);

            ArrayList<PhotoEntity> photos = new ArrayList<>(safeCountOf(cursor));
            if (nonNull(cursor)) {
                while (cursor.moveToNext()) {
                    if (e.isDisposed()) {
                        break;
                    }

                    photos.add(mapPhotoDbo(cursor));
                }

                cursor.close();
            }

            e.onSuccess(photos);
        });
    }

    @Override
    public Completable applyPatch(int accountId, int ownerId, int photoId, PhotoPatch patch) {
        return Completable.fromAction(() -> {
            ContentValues cv = new ContentValues();

            if (nonNull(patch.getLike())) {
                cv.put(PhotosColumns.LIKES, patch.getLike().getCount());
                cv.put(PhotosColumns.USER_LIKES, patch.getLike().isLiked());
            }

            if (nonNull(patch.getDeletion())) {
                cv.put(PhotosColumns.DELETED, patch.getDeletion().isDeleted());
            }

            if (cv.size() > 0) {
                Uri uri = MessengerContentProvider.getPhotosContentUriFor(accountId);
                String where = PhotosColumns.PHOTO_ID + " = ? AND " + PhotosColumns.OWNER_ID + " = ?";
                String[] args = {String.valueOf(photoId), String.valueOf(ownerId)};

                getContentResolver().update(uri, cv, where, args);
            }
        });
    }

    private PhotoEntity mapPhotoDbo(Cursor cursor) {
        PhotoSizeEntity sizes = null;

        String sizesJson = cursor.getString(cursor.getColumnIndex(PhotosColumns.SIZES));
        if (nonEmpty(sizesJson)) {
            sizes = GSON.fromJson(sizesJson, PhotoSizeEntity.class);
        }

        int id = cursor.getInt(cursor.getColumnIndex(PhotosColumns.PHOTO_ID));
        int ownerId = cursor.getInt(cursor.getColumnIndex(PhotosColumns.OWNER_ID));
        return new PhotoEntity().set(id, ownerId)
                .setSizes(sizes)
                .setAlbumId(cursor.getInt(cursor.getColumnIndex(PhotosColumns.ALBUM_ID)))
                .setWidth(cursor.getInt(cursor.getColumnIndex(PhotosColumns.WIDTH)))
                .setHeight(cursor.getInt(cursor.getColumnIndex(PhotosColumns.HEIGHT)))
                .setText(cursor.getString(cursor.getColumnIndex(PhotosColumns.TEXT)))
                .setDate(cursor.getLong(cursor.getColumnIndex(PhotosColumns.DATE)))
                .setUserLikes(cursor.getInt(cursor.getColumnIndex(PhotosColumns.USER_LIKES)) == 1)
                .setCanComment(cursor.getInt(cursor.getColumnIndex(PhotosColumns.CAN_COMMENT)) == 1)
                .setLikesCount(cursor.getInt(cursor.getColumnIndex(PhotosColumns.LIKES)))
                .setCommentsCount(cursor.getInt(cursor.getColumnIndex(PhotosColumns.COMMENTS)))
                .setTagsCount(cursor.getInt(cursor.getColumnIndex(PhotosColumns.TAGS)))
                .setAccessKey(cursor.getString(cursor.getColumnIndex(PhotosColumns.ACCESS_KEY)))
                .setDeleted(cursor.getInt(cursor.getColumnIndex(PhotosColumns.DELETED)) == 1);
    }
}

File: app/src/main/java/dev/ragnarok/fenrir/db/impl/PhotosStorage.kt
package dev.ragnarok.fenrir.db.impl

import android.content.ContentProviderOperation
import android.content.ContentValues
import android.database.Cursor
import android.provider.BaseColumns
import dev.ragnarok.fenrir.db.MessengerContentProvider
import dev.ragnarok.fenrir.db.MessengerContentProvider.Companion.getPhotosContentUriFor
import dev.ragnarok.fenrir.db.column.PhotosColumns
import dev.ragnarok.fenrir.db.interfaces.IPhotosStorage
import dev.ragnarok.fenrir.db.model.PhotoPatch
import dev.ragnarok.fenrir.db.model.entity.PhotoEntity
import dev.ragnarok.fenrir.db.model.entity.PhotoSizeEntity
import dev.ragnarok.fenrir.model.criteria.PhotoCriteria
import dev.ragnarok.fenrir.nonNullNoEmpty
import dev.ragnarok.fenrir.util.Utils.safeCountOf
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Single
import io.reactivex.rxjava3.core.SingleEmitter

internal class PhotosStorage(base: AppStorages) : AbsStorage(base), IPhotosStorage {
    override fun insertPhotosRx(
        accountId: Int,
        ownerId: Int,
        albumId: Int,
        photos: List<PhotoEntity>,
        clearBefore: Boolean
    ): Completable {
        return Completable.fromAction {
            val operations =
                ArrayList<ContentProviderOperation>(if (clearBefore) photos.size + 1 else photos.size)
            val uri = getPhotosContentUriFor(accountId)
            if (clearBefore) {
                operations.add(
                    ContentProviderOperation
                        .newDelete(uri)
                        .withSelection(
                            PhotosColumns.OWNER_ID + " = ? AND " + PhotosColumns.ALBUM_ID + " = ?",
                            arrayOf(ownerId.toString(), albumId.toString())
                        )
                        .build()
                )
            }
            for (dbo in photos) {
                operations.add(
                    ContentProviderOperation
                        .newInsert(uri)
                        .withValues(getCV(dbo))
                        .build()
                )
            }
            context.contentResolver.applyBatch(MessengerContentProvider.AUTHORITY, operations)
        }
    }

    override fun findPhotosByCriteriaRx(criteria: PhotoCriteria): Single<List<PhotoEntity>> {
        return Single.create { e: SingleEmitter<List<PhotoEntity>> ->
            val selection = getSelectionForCriteria(criteria)
            val orderBy =
                if (criteria.orderBy == null) PhotosColumns.PHOTO_ID + (if (!criteria.sortInvert) " DESC" else " ASC") else criteria.orderBy
            val uri = getPhotosContentUriFor(criteria.accountId)
            val cursor = context.contentResolver.query(uri, null, selection, null, orderBy)
            val photos = ArrayList<PhotoEntity>(safeCountOf(cursor))
            if (cursor != null) {
                while (cursor.moveToNext()) {
                    if (e.isDisposed) {
                        break
                    }
                    photos.add(mapPhotoDbo(cursor))
                }
                cursor.close()
            }
            e.onSuccess(photos)
        }
    }

    override fun applyPatch(
        accountId: Int,
        ownerId: Int,
        photoId: Int,
        patch: PhotoPatch
    ): Completable {
        return Completable.fromAction {
            val cv = ContentValues()
            if (patch.like != null) {
                cv.put(PhotosColumns.LIKES, patch.like.count)
                cv.put(PhotosColumns.USER_LIKES, patch.like.isLiked)
            }
            if (patch.deletion != null) {
                cv.put(PhotosColumns.DELETED, patch.deletion.isDeleted)
            }
            if (cv.size() > 0) {
                val uri = getPhotosContentUriFor(accountId)
                val where = PhotosColumns.PHOTO_ID + " = ? AND " + PhotosColumns.OWNER_ID + " = ?"
                val args = arrayOf(photoId.toString(), ownerId.toString())
                contentResolver.update(uri, cv, where, args)
            }
        }
    }

    private fun mapPhotoDbo(cursor: Cursor): PhotoEntity {
        var sizes: PhotoSizeEntity? = null
        val sizesJson = cursor.getString(cursor.getColumnIndexOrThrow(PhotosColumns.SIZES))
        if (sizesJson.nonNullNoEmpty()) {
            sizes = GSON.fromJson(sizesJson, PhotoSizeEntity::class.java)
        }
        val id = cursor.getInt(cursor.getColumnIndexOrThrow(PhotosColumns.PHOTO_ID))
        val ownerId = cursor.getInt(cursor.getColumnIndexOrThrow(PhotosColumns.OWNER_ID))
        return PhotoEntity().set(id, ownerId)
            .setSizes(sizes)
            .setAlbumId(cursor.getInt(cursor.getColumnIndexOrThrow(PhotosColumns.ALBUM_ID)))
            .setWidth(cursor.getInt(cursor.getColumnIndexOrThrow(PhotosColumns.WIDTH)))
            .setHeight(cursor.getInt(cursor.getColumnIndexOrThrow(PhotosColumns.HEIGHT)))
            .setText(cursor.getString(cursor.getColumnIndexOrThrow(PhotosColumns.TEXT)))
            .setDate(cursor.getLong(cursor.getColumnIndexOrThrow(PhotosColumns.DATE)))
            .setUserLikes(cursor.getInt(cursor.getColumnIndexOrThrow(PhotosColumns.USER_LIKES)) == 1)
            .setCanComment(cursor.getInt(cursor.getColumnIndexOrThrow(PhotosColumns.CAN_COMMENT)) == 1)
            .setLikesCount(cursor.getInt(cursor.getColumnIndexOrThrow(PhotosColumns.LIKES)))
            .setCommentsCount(cursor.getInt(cursor.getColumnIndexOrThrow(PhotosColumns.COMMENTS)))
            .setTagsCount(cursor.getInt(cursor.getColumnIndexOrThrow(PhotosColumns.TAGS)))
            .setAccessKey(cursor.getString(cursor.getColumnIndexOrThrow(PhotosColumns.ACCESS_KEY)))
            .setDeleted(cursor.getInt(cursor.getColumnIndexOrThrow(PhotosColumns.DELETED)) == 1)
    }

    companion object {
        private fun getCV(dbo: PhotoEntity): ContentValues {
            val cv = ContentValues()
            cv.put(PhotosColumns.PHOTO_ID, dbo.id)
            cv.put(PhotosColumns.ALBUM_ID, dbo.albumId)
            cv.put(PhotosColumns.OWNER_ID, dbo.ownerId)
            cv.put(PhotosColumns.WIDTH, dbo.width)
            cv.put(PhotosColumns.HEIGHT, dbo.height)
            cv.put(PhotosColumns.TEXT, dbo.text)
            cv.put(PhotosColumns.DATE, dbo.date)
            if (dbo.sizes != null) {
                cv.put(PhotosColumns.SIZES, GSON.toJson(dbo.sizes))
            }
            cv.put(PhotosColumns.USER_LIKES, dbo.isUserLikes)
            cv.put(PhotosColumns.CAN_COMMENT, dbo.isCanComment)
            cv.put(PhotosColumns.LIKES, dbo.likesCount)
            cv.put(PhotosColumns.COMMENTS, dbo.commentsCount)
            cv.put(PhotosColumns.TAGS, dbo.tagsCount)
            cv.put(PhotosColumns.ACCESS_KEY, dbo.accessKey)
            cv.put(PhotosColumns.DELETED, dbo.isDeleted)
            return cv
        }

        private fun getSelectionForCriteria(criteria: PhotoCriteria): String {
            var selection = "1 = 1"
            if (criteria.ownerId != null) {
                selection = selection + " AND " + PhotosColumns.OWNER_ID + " = " + criteria.ownerId
            }
            if (criteria.albumId != null) {
                selection = selection + " AND " + PhotosColumns.ALBUM_ID + " = " + criteria.albumId
            }
            val range = criteria.range
            if (range != null) {
                selection = selection + " AND " + BaseColumns._ID + " >= " + range.first +
                        " AND " + BaseColumns._ID + " <= " + criteria.range.last
            }
            return selection
        }
    }
}